querier.hh

```diff
@@ -146,7 +155,7 @@ public:
         , _permit(std::move(permit))
         , _range(std::make_unique<const dht::partition_range>(std::move(range)))
         , _slice(std::make_unique<const query::partition_slice>(std::move(slice)))
-        , _reader(ms.make_reader(_schema, _permit, *_range, *_slice, pc, std::move(trace_ptr), streamed_mutation::forwarding::no, mutation_reader::forwarding::no))
+        , _reader(ms.make_reader(underlying_schema(), _permit, *_range, *_slice, pc, std::move(trace_ptr), streamed_mutation::forwarding::no, mutation_reader::forwarding::no))
```

Comment:
To be correct, we should also pass a non-reversed _slice. Otherwise, if one of the downstream readers is converted to the new API, it will read in reverse using non-reversed schema.

Reply:
Also, a non-reversing reader will expect non-reversed clustering ranges.

Reply:
\cc @enedil

Reply:
We have three pieces of information which need to be in sync.

1. schema - are we using the query schema or the table schema
2. slice - are we using the query slice or the table slice
3. slice::options::reversed - are we explicitly asking for reversal, or do we let the schema's domain to make the decision

I think that we need to either pass query-domain schema+slice, OR set the reversed bit, but not both.

Passing a reversed schema and slice to the reader says: here are mutually-consistent schema and slice, give me fragments in natural order. It is up to you, the reader [1], to figure out that the schema may be reversed wrt. the table schema and do something about that if you need to.

Passing a option_set::reversed_bit and a table-domain schema and slice says: here is a query in the table domain. Perform it for me, but then return the results backwards.

Passing both a reversed schema and the reversed bit means double reversal and should not be done.

Perhaps we can have a make_flat_mutation_reader assist as with this.
1. check whether Impl::native_reversed_query is true
2. if it is, look for a slice parameter, check if the reversed bit is set. If it is, allocate a new slice parameter (and stow it together with Impl) with the reversed bit turned off and the intervals reversed, and reverse any schema parameter
3. if not, then look for a schema parameter, check if it's reversed. If it is, allocate a new slice parameter (and stow it together with Impl) with the reversed bit turned on and the intervals (re/un)-reversed, and (re/un)-reverse any schema parameter

This moves all the glue into make_flat_mutation_reader, provided it's the only way to create flat mutation readers. Later we can drop the new logic after everything is converted.


[1] "reader" here is the flat mutation reader, not you

Reply:
> To be correct, we should also pass a non-reversed _slice. Otherwise, if one of the downstream readers is converted to the new API, it will read in reverse using non-reversed schema.

The current sstable reader expects a non-reversed schema and a reversed slice. Once we merge the reversing sstable reader it will expect reversed schema and reversed slice. I don't think it is worth reversing the slice here (and reversing it again in the sstable reader) for just this small window.

Reply:
@avikivity this will make the generic reader aware of reverse queries which is what we wanted to avoid by the reverse schema shenanigans. I think we can work with much simpler rules:
1. schema - reversed in all layers except the mutation sources
2. slice - always reversed
3. slice::options::reversed - always set

We have two kind of layers in the code:
1. Unaware of reverse read: they never check the reverse bit and don't need to, they will have a schema matching the clustering ranges in the slice
2. Aware of revers reads: they will work with two schemas: native and read schema and know how to handle the slice

Reply:
So, the slice is always in the query domain? Doesn't that mean that this series has to convert it back to the table domain for all the readers that work in the table domain?

For sure we have more than one reader that works in the table domain. The cache and memtable readers, for example.

Reply:
Yes, and readers that do slicing already work with it that way, through `clustering_key_filter_ranges`, which is aware of reverse slices and abstracts all this away.

Reply:
I just discovered that we don't fully reverse slices, we only reverse the ranges, but not the range bounds themselves.

Reply:
Ah, I guess you mean that slices are already in the query domain even before the series. But they're not fully in the query domain because the bounds are reversed. I guess a patch that fixes that (and touches all readers that care) can be applied to remove the problem (we'll still have the half-reversal on the wire).

Reply:
Yes.

Reply:
Ok. I can review such a patch ahead of this series, or as part of it.

Reply:
I think it would best be part of this series.